app/iOSFS.m

```diff
@@ -0,0 +1,320 @@
+//
+//  ExternalFolder.m
+//  iSH
+//
+//  Created by Noah Peeters on 26.10.19.
+//
+
+#import <Foundation/Foundation.h>
+#import <UIKit/UIKit.h>
+#include "AppDelegate.h"
+#include "iOSFS.h"
+#include "kernel/fs.h"
+#include "kernel/errno.h"
+#include "fs/path.h"
+
+const NSFileCoordinatorWritingOptions NSFileCoordinatorWritingForCreating = NSFileCoordinatorWritingForMerging;
+
+@interface DirectoryPickerDelegate : NSObject <UIDocumentPickerDelegate>
+
+@property NSURL *url;
+@property pthread_mutex_t *mutex;
+
+@end
+
+@implementation DirectoryPickerDelegate
+
+- (instancetype)initWithMutex:(pthread_mutex_t *)mutex {
+    if (self = [super init]) {
+        self.mutex = mutex;
+        self.url = nil;
+        pthread_mutex_lock(mutex);
+    }
+    return self;
+}
+
+- (void)documentPickerWasCancelled:(UIDocumentPickerViewController *)controller {
+    pthread_mutex_unlock(self.mutex);
+}
+
+- (void)documentPicker:(UIDocumentPickerViewController *)controller didPickDocumentsAtURLs:(NSArray<NSURL *> *)urls {
+    if (urls.count > 0)
+        self.url = urls[0];
+    pthread_mutex_unlock(self.mutex);
+}
+
+@end
+
+NSURL *url_for_mount(struct mount *mount) {
+    return (__bridge NSURL *) mount->data;
+}
+
+NSURL *url_for_path_in_mount(struct mount *mount, const char *path) {
+    return [url_for_mount(mount) URLByAppendingPathComponent:[NSString stringWithUTF8String:path]];
+}
+
+static int combine_error(NSError *coordinatorError, int err) {
+    if (coordinatorError != nil)
+        return _EPERM;
+
+    return err;
+}
+
+static struct fd *iosfs_open(struct mount *mount, const char *path, int flags, int mode) {
+    return realfs.open(mount, path, flags, mode);
+}
+
+static int iosfs_mount_data(char source[MAX_PATH], char point[MAX_PATH], dword_t *flags, void **data) {
+    pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
```

Comment:
Can this be moved into iosfs_mount? I think that would eliminate the need for a new do_mount_with_data.